### ISSUE TITLE
Tree Shaking ability in Vue

### ISSUE THREAD

**Issue by paritosh149** [vuejs/vue-cli#1644]
### What problem does this feature solve?
Tree shaking would help in reducing the output file size and improve overall performance.

### What does the proposed API look like?
vue-cli-service build --optimize

**Comment by LinusBorg** [comment]
Webpack does tree shaking out of the box, I'm not sure what else you expect.

**Comment by paritosh149** [comment]
I used latest vue-cli to create and build standard template app. Inside dist/js folder, the size of vendor.js is 120,162 bytes, unzipped. I might sound naive, but isn't this too much size for an app that simple. This query is because of Vue app's size comparison to Angular's Ivy project which achieved total size of its Hello World app to approx 10KB.

**Comment by uoziod** [comment]
@paritosh149, this is very good point. I really think this is something to consider in further development.

**Comment by LinusBorg** [comment]
Well, install webpack-bundle-analyzer and have a look for yourself to see what's going on. I don't know what you did your did not include in your build.

Also, a few notes:

* Ivy's small footprint for helloworld is impressive, but since we never serve "helloWorld" to our customers, it's not a very good benchmark for anything. We should look at an app that includes stuff apps use in 99% of cases and see what size it compiles to then.
* vue-core itsef is ~20kb gzipped, vue-cli can't do much about that. If your request was aimed to slim down the core itself with "tree shaking", you are asking in the wrong repository. What I can tell you is hat we will have a more modular approach for Vue 3 that minght enable some improvements in that direction.
* vue-cli 3 includes some basic polyfills b default that apps will need in 99% of cases - that does increase bundle size

**Comment by uoziod** [comment]
> Ivy's small footprint for helloworld is impressive, but since we never serve "helloWorld" to our customers, it's not a very good benchmark for anything. We should look at an app that includes stuff apps use in 99% of cases and see what size it compiles to then.

It is very good benchmark to clearly see how good framework code susceptible to treeshaking.

**Comment by LinusBorg** [comment]
Well, vue-core doesn't do/support any tree-shaking right now. We can surely discuss how that can be changed in the future, but as mentioned, that's not related to vue-cli

**Comment by bberneker-ikea** [comment]
Even if vue cli doesn't provide the kind of tree shaking requested out of the box, does anyone have some resources they can share on how to do this through manual configuration of webpack?  I have a project I'd like to reduce and don't mind a little extra configuration effort if I can just find an understandable guide.

**Comment by LinusBorg** [comment]
> does anyone have some resources they can share on how to do this through manual configuration of webpack?

There's no way to achieve that with Vue 2, plain and simple.

Vue 3 will be designed with tree-shakability in mind.

**Comment by bberneker-ikea** [comment]
Perhaps my request is out of scope for this thread.
What I'm asking for is not to tree shake vue itself, but the various modules and dependencies I am using in a project created using the vue cli.  Can't I manually tweak webpack to tree shake my imports?

**Comment by LinusBorg** [comment]
That's really out of scope. 

Come to our forum at forurm.vuejs.orf

**Comment by paritosh149** [comment]
> That's really out of scope.
> 
> Come to our forum at forurm.vuejs.org

Typo Correction -> the link is [forum.vuejs.org](forum.vuejs.org)

**Comment by uptownhr** [comment]
@LinusBorg I know you've closed this issue a while back, so I'm sorry for pinging you here. But I've tried searching google and the forum and wasn't able to find a follow up regarding this thread. 

But I'd like to ask if "multiple export files" (not sure if this is the right word) would be supported. The question of tree shaking is when the output produced by `vue-cli-service build` can be tree shaken based on what is consumed by another project. 

For example, I have a UI library that I've produced. In a target project, I've installed the UI library and but only use a single component. The result of the import given the single `main.js` entrypoint results in the entire library being imported and built. 

Given that `vue-cli-service` supports the `wc` mode, I'm suprised the `lc` method wouldn't also create builds the same way or an option to do so. 

Not sure if closed issues will get to you but this would be awesome feature added to support  component development.
